packages/core/src/hooks/form/index.ts

```diff
@@ -1 +1,341 @@
+ * @typeParam TVariables - Values for params. default `{}`
+ * @typeParam TData - Result data returned by the `select` function. Extends {@link https://refine.dev/docs/core/interface-references/#baserecord `BaseRecord`}. Defaults to `TQueryFnData`
+ * @typeParam TResponse - Result data returned by the mutation function. Extends {@link https://refine.dev/docs/core/interface-references/#baserecord `BaseRecord`}. Defaults to `TData`
+ * @typeParam TResponseError - Custom error object that extends {@link https://refine.dev/docs/core/interface-references/#httperror `HttpError`}. Defaults to `TError`
+ *
+ */
+export const useForm = <
+  TQueryFnData extends BaseRecord = BaseRecord,
+  TError extends HttpError = HttpError,
+  TVariables = {},
+  TData extends BaseRecord = TQueryFnData,
+  TResponse extends BaseRecord = TData,
+  TResponseError extends HttpError = TError,
+>(
+  props: UseFormProps<
+    TQueryFnData,
+    TError,
+    TVariables,
+    TData,
+    TResponse,
+    TResponseError
+  > = {},
+): UseFormReturnType<
+  TQueryFnData,
+  TError,
+  TVariables,
+  TData,
+  TResponse,
+  TResponseError
+> => {
+  const getMeta = useMeta();
+  const invalidate = useInvalidate();
+  const { redirect: defaultRedirect } = useRefineOptions();
+  const { mutationMode: defaultMutationMode } = useMutationMode();
+
+  const { setWarnWhen } = useWarnAboutChange();
+  const handleSubmitWithRedirect = useRedirectionAfterSubmission();
+
+  const pickedMeta = pickNotDeprecated(props.meta, props.metaData);
+  const mutationMode = props.mutationMode ?? defaultMutationMode;
+
+  const {
+    resource,
+    action: routeAction,
+    identifier,
+  } = useResource(props.resource);
+
+  const [autosaved, setAutosaved] = React.useState(false);
+
+  const [id, setId] = useId({ id: props.id, resource: props.resource });
+
+  const action = (() => {
+    // if action is provided from props, use it
+    if (props.action) return props.action;
+    // fallback to create if action is not edit or clone
+    if (routeAction !== "edit" && routeAction !== "clone") return "create";
+    // otherwise use the route action
+    return routeAction;
+  })();
+
+  const isEdit = action === "edit";
+  const isClone = action === "clone";
+  const isCreate = action === "create";
+
+  const combinedMeta = getMeta({
+    resource,
+    meta: pickedMeta,
+  });
+
+  const idRequired = (isEdit || isClone) && Boolean(props.resource);
+  const idHandled = typeof props.id !== "undefined";
+  const idDismissed = props.queryOptions?.enabled === false;
+
+  /**
+   * When a custom resource is provided through props, `id` will not be inferred from the URL to avoid any potential faulty requests.
+   * In this case, `id` is required to be passed through props.
+   * If `id` is not handled, a warning will be thrown in development mode.
+   */
+  warnOnce(
+    idRequired && !idHandled && !idDismissed,
```

Comment:
```suggestion
    idRequired && !hasIdProp && !isQueryDisabled,
```

With the suggested changes above, I think this condition becomes easier to read.

Reply:
I think we can also use `id` variable here, coming from `useId` hook.